### PR TITLE
Add AgriFieldNet India Challenge dataset

### PR DESCRIPTION
**Pull request by yichiac**
This PR adds the AgriFieldNet India Challenge dataset. The dataset contains pre-chipped Sentinel-2 imagery and labels for crop type in India. 

[Challenge](https://zindi.africa/competitions/agrifieldnet-india-challenge)
[Dataset overview](https://beta.source.coop/radiantearth/agrifieldnet-competition/)

The downloader for Source Cooperative needs to migrate from Radiant MLHub to Source Cooperative #1830. Due to the migration process not yet being finalized, the PR currently does not implement the automatic download feature.

A significant portion of the datasets available through this service are provided as compressed files, which facilitates the use of checksums for integrity verification. However, this dataset is distributed via directories and individual files. This structure presents a challenge for implementing standard checksum verification procedures.

Given this scenario, we have yet to establish a robust method for checksumming these non-compressed files and directories. Consequently, this PR temporarily suspends the execution of `check_integrity` and related checksum functions until a viable solution is determined.

agrifieldnet_test
agrifieldnet_test2

### CODE REVIEW

**Comment by adamjstewart** [comment]
Not sure if you can see it but https://app.codecov.io/gh/microsoft/torchgeo/pull/1459 tells you which lines are missing coverage. You can also check this from the command line: https://torchgeo.readthedocs.io/en/stable/user/contributing.html#tests

File: torchgeo/datamodules/agrifieldnet.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""DeepGlobe Land Cover Classification Challenge datamodule."""
```
Comment (adamjstewart) [inline torchgeo/datamodules/agrifieldnet.py:4]:
```suggestion
"""AgriFieldNet datamodule."""
```

File: torchgeo/datamodules/agrifieldnet.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+from ..datasets import AgriFieldNet
+from ..samplers.utils import _to_tuple
+from ..transforms import AugmentationSequential
+from ..transforms.transforms import _RandomNCrop
+from .geo import NonGeoDataModule
+from .utils import dataset_split
+
+
+class AgriFieldNetDataModule(NonGeoDataModule):
+    """LightningDataModule implementation for the DeepGlobe Land Cover dataset.
```
Comment (adamjstewart) [inline torchgeo/datamodules/agrifieldnet.py:19]:
```suggestion
    """LightningDataModule implementation for the AgriFieldNet dataset.
```

File: torchgeo/datamodules/agrifieldnet.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    def __init__(
+        self,
+        batch_size: int = 64,
+        patch_size: Union[tuple[int, int], int] = 256,
+        val_split_pct: float = 0.2,
+        num_workers: int = 0,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize a new DeepGlobeLandCoverDataModule instance.
```
Comment (adamjstewart) [inline torchgeo/datamodules/agrifieldnet.py:32]:
```suggestion
        """Initialize a new AgriFieldNetDataModule instance.
```

File: torchgeo/datamodules/agrifieldnet.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+        Args:
+            batch_size: Size of each mini-batch.
+            patch_size: Size of each patch, either ``size`` or ``(height, width)``.
+                Should be a multiple of 32 for most segmentation architectures.
+            val_split_pct: Percentage of the dataset to use as a validation set.
+            num_workers: Number of workers for parallel data loading.
+            **kwargs: Additional keyword arguments passed to
+                :class:`~torchgeo.datasets.DeepGlobeLandCover`.
```
Comment (adamjstewart) [inline torchgeo/datamodules/agrifieldnet.py:41]:
```suggestion
                :class:`~torchgeo.datasets.AgriFieldNet`.
```

File: torchgeo/datamodules/agrifieldnet.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        """
+        super().__init__(AgriFieldNet, 1, num_workers, **kwargs)
+
+        self.patch_size = _to_tuple(patch_size)
+        self.val_split_pct = val_split_pct
+
+        self.aug = AugmentationSequential(
+            K.Normalize(mean=self.mean, std=self.std),
+            _RandomNCrop(self.patch_size, batch_size),
```
Comment (adamjstewart) [inline torchgeo/datamodules/agrifieldnet.py:50]: If the tiles are already 256x256 chips there might be no need to crop them, and cropping them N times definitely doesn't make sense. This is designed for very large tiles

File: torchgeo/datasets/agrifieldnet.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+    9 - Maize
+    13 - Gram
+    14 - Coriander
+    15 - Potato
+    16 - Bersem
+    36 - Rice
+
+    If you use this dataset in your research, please cite the following dataset:
+        Radiant Earth Foundation & IDinsight (2022) AgriFieldNet Competition Dataset,
+        Version 1.0, Radiant MLHub. https://doi.org/10.34911/rdnt.wu92p1
```
Comment (adamjstewart) [inline torchgeo/datasets/agrifieldnet.py:69]:
```suggestion

    * https://doi.org/10.34911/rdnt.wu92p1
```

Comment (adamjstewart) [reply torchgeo/datasets/agrifieldnet.py:69]: Is the MLHub website down? I can't access it

File: torchgeo/datasets/agrifieldnet.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+    filename = "ref_agrifieldnet_competition_v1.tar.gz"
+    md5 = "85055da1e7eb69fa4b3d925ee1450a74"
+    splits = ["train", "test"]
+    collections = [
+        "ref_agrifieldnet_competition_v1_source",
+        "ref_agrifieldnet_competition_v1_labels_train",
+        "ref_agrifieldnet_competition_v1_labels_test",
+    ]
+
+    # missing class corresponding values
```
Comment (adamjstewart) [inline torchgeo/datasets/agrifieldnet.py:82]: What does this comment mean?

**Review by adamjstewart — COMMENTED** [review]
Can you upload an example plot to make sure everything looks correct?

File: torchgeo/datamodules/agrifieldnet.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+from ..datasets import AgriFieldNet
+from ..samplers.utils import _to_tuple
+from ..transforms import AugmentationSequential
+from ..transforms.transforms import _RandomNCrop
+from .geo import NonGeoDataModule
+from .utils import dataset_split
+
+
+class AgriFieldNetDataModule(NonGeoDataModule):
```
Comment (adamjstewart) [inline torchgeo/datamodules/agrifieldnet.py:18]: Needs versionadded docs

**Comment by calebrob6** [comment]
@yichiac do you need help with this?

**Comment by yichiac** [comment]
> @yichiac do you need help with this?

**Comment by yichiac** [comment]
> @yichiac do you need help with this?

The previous version still needs to complete codecov test for the datamodule (other tests are mostly done). I keep using this branch for developing other features, such as generating submission files for the [contest](https://zindi.africa/competitions/agrifieldnet-india-challenge). I've noticed that a small portion of the test results are missing, but I'm actively fixing this issue. I may need help to complete the codecov test for the datamodule. Once that is finished, I can add more features based on the dataset on torchgeo.

**Comment by adamjstewart** [comment]
Is this still a WIP? There seem to be a few training script things that are out of date.

**Comment by yichiac** [comment]
Yes, it's WIP. The out of date code will be fixed

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Let's remove everything related to downloading (including comments) until we actually support it in the future.

File: torchgeo/datasets/agrifieldnet.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+    The `AgriFieldNet India Challenge
+    <https://zindi.africa/competitions/agrifieldnet-india-challenge>`__ dataset
+    includes satellite imagery from Sentinel-2 cloud free composites
+    (single snapshot) and labels for crop type that were collected by ground survey.
+    The Sentinel-2 data are then matched with corresponding labels.
+    The dataset contains 7081 fields, which have been split into training and
+    test sets (5551 fields in the train and 1530 fields in the test).
+    Satellite imagery and labels are tiled into 256x256 chips adding up to 1217 tiles.
+    The fields are distributed across all chips, some chips may only have train or
+    test fields and some may have both. Since the labels are derived from data
```
[the remark this answers is not in the document]
Comment (yichiac) [reply torchgeo/datasets/agrifieldnet.py:37]: Yes, it's an interesting split method. The test labels are not publicly available and can only be tested by submitting them to the competition website. The files in test_labels are only field_id masks. 

For this competition, the submission form should be a CSV file including test_field_ids and corresponding class prediction probabilities. 

Splitting label files in the train_label folder should be enough to train models to measure spatial generalizability.

In this case, is it still necessary to add a `split` parameter in this PR?

[the remark this answers is not in the document]
Comment (adamjstewart) [reply torchgeo/datasets/agrifieldnet.py:33]: Ah, I see. So only the train set is public. So for our data module, we'll have to split train into train-val-test. Hopefully we have enough data for that. In that case, no need to add `split`, we'll do all splitting in the data module.

File: torchgeo/datasets/agrifieldnet.py [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+    of ground reference data which is a skewed distribution with a few dominant
+    crops being over represented. The test set was drawn randomly from an area
+    weighted field list that ensured that fields with less common crop types
+    were better represented in the test set. The original dataset can be
+    downloaded from `Source Cooperative <https://beta.source.coop/
+    radiantearth/agrifieldnet-competition/>`__.
+
+    Dataset format:
+
+    * images are 12-bands Sentinel-2 data
```
Comment (adamjstewart) [inline torchgeo/datasets/agrifieldnet.py:48]:
```suggestion
    * images are 12-band Sentinel-2 data
```

File: torchgeo/datasets/agrifieldnet.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+    * masks are tiff image with unique values representing the class and field id
```
Comment (adamjstewart) [inline torchgeo/datasets/agrifieldnet.py:49]:
```suggestion
    * masks are tiff images with unique values representing the class and field id
```

File: torchgeo/datasets/agrifieldnet.py [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+    0 - No-Data
+    1 - Wheat
+    2 - Mustard
+    3 - Lentil
+    4 - No Crop/Fallow
+    5 - Green pea
+    6 - Sugarcane
+    8 - Garlic
+    9 - Maize
+    13 - Gram
```
Comment (adamjstewart) [inline torchgeo/datasets/agrifieldnet.py:62]: Didn't know this was another name for chickpea, interesting.

File: torchgeo/datasets/agrifieldnet.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+    14 - Coriander
+    15 - Potato
+    16 - Bersem
```
Comment (adamjstewart) [inline torchgeo/datasets/agrifieldnet.py:65]: Should this be Berseem?

File: torchgeo/datasets/agrifieldnet.py [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+    36 - Rice
+
+    If you use this dataset in your research, please cite the following dataset:
+
+    * https://doi.org/10.34911/rdnt.wu92p1
+
+    .. versionadded:: 0.6
+    """
+
+    url = "https://radiantearth.blob.core.windows.net/mlhub/ref_agrifieldnet_competition_v1"  # noqa: E501
```
Comment (adamjstewart) [inline torchgeo/datasets/agrifieldnet.py:75]: Remove

File: torchgeo/datasets/agrifieldnet.py [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+
+        hits = self.index.intersection(tuple(query), objects=True)
+        filepaths = cast(list[str], [hit.object for hit in hits])
+
+        if not filepaths:
+            raise IndexError(
+                f"query: {query} not found in index with bounds: {self.bounds}"
+            )
+
+        if self.separate_files:
```
Comment (adamjstewart) [inline torchgeo/datasets/agrifieldnet.py:180]: No need for this if-statement, it's always true

**Review by adamjstewart — COMMENTED** [review]
Mostly good, just a couple remaining comments

File: torchgeo/datasets/agrifieldnet.py (hunk 13 above)
Comment (yichiac) [reply torchgeo/datasets/agrifieldnet.py:65]: Berseem is more reasonable. The [competition website](https://zindi.africa/competitions/agrifieldnet-india-challenge/data) and [GitHub](https://github.com/radiantearth/agrifieldnet_india_competition) both said Bersem. But I can't find much information about Bersem. I prefer Berseem for now.

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
Confirmed 100% test coverage locally.
